internal/crypto/cert.go

```diff
@@ -50,124 +71,10 @@ func CNFromDeviceFingerprint(fingerprint string) (string, error) {
-func MakeSelfSignedCA(certFile, keyFile, serialFile, subjectName string, expiryDays int) (*CA, error) {
-	caConfig, err := makeSelfSignedCAConfig(
-		pkix.Name{CommonName: subjectName},
-		time.Duration(expiryDays)*24*time.Hour,
-	)
-	if err != nil {
-		return nil, err
-	}
-	if err = caConfig.WriteCertConfigFile(certFile, keyFile); err != nil {
-		return nil, err
-	}
-
-	var serialGenerator oscrypto.SerialGenerator
-	if len(serialFile) > 0 {
-		// create / overwrite the serial file with a zero padded hex value (ending in a newline to have a valid file)
-		if err := os.WriteFile(serialFile, []byte("00\n"), 0600); err != nil {
-			return nil, err
-		}
-		serialGenerator, err = oscrypto.NewSerialFileGenerator(serialFile)
-		if err != nil {
-			return nil, err
-		}
-	} else {
-		serialGenerator = &oscrypto.RandomSerialGenerator{}
-	}
-
-	config := TLSCertificateConfig(*caConfig)
-	return &CA{
-		SerialGenerator: serialGenerator,
-		Config:          &config,
-	}, nil
-}
-
-func makeSelfSignedCAConfig(subject pkix.Name, caLifetime time.Duration) (*oscrypto.TLSCertificateConfig, error) {
-	rootcaPublicKey, rootcaPrivateKey, publicKeyHash, err := NewKeyPairWithHash()
-	if err != nil {
-		return nil, err
-	}
-
-	now := time.Now()
-	rootcaTemplate := &x509.Certificate{
-		Subject: subject,
-
-		SignatureAlgorithm: x509.ECDSAWithSHA256,
-
-		NotBefore: now.Add(-1 * time.Second),
-		NotAfter:  now.Add(caLifetime),
-
-		SerialNumber: randomSerial(),
-
-		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
-		BasicConstraintsValid: true,
-		IsCA:                  true,
-
-		AuthorityKeyId: publicKeyHash,
-		SubjectKeyId:   publicKeyHash,
-	}
-	rootcaCert, err := signCertificate(rootcaTemplate, rootcaPublicKey, rootcaTemplate, rootcaPrivateKey)
-	if err != nil {
-		return nil, err
-	}
-	caConfig := &oscrypto.TLSCertificateConfig{
-		Certs: []*x509.Certificate{rootcaCert},
-		Key:   rootcaPrivateKey,
-	}
-	return caConfig, nil
-}
-
-func signCertificate(template *x509.Certificate, requestKey crypto.PublicKey, issuer *x509.Certificate, issuerKey crypto.PrivateKey) (*x509.Certificate, error) {
-	derBytes, err := x509.CreateCertificate(rand.Reader, template, issuer, requestKey, issuerKey)
-	if err != nil {
-		return nil, err
-	}
-	certs, err := x509.ParseCertificates(derBytes)
-	if err != nil {
-		return nil, err
-	}
-	if len(certs) != 1 {
-		return nil, errors.New("expected a single certificate")
-	}
-	return certs[0], nil
-}
-
-// func (ca *CA) EnsureSubCA(certFile, keyFile, serialFile, name string, expireDays int) (*CA, bool, error) {
-// 	if subCA, err := GetCA(certFile, keyFile, serialFile); err == nil {
-// 		return subCA, false, err
-// 	}
-// 	subCA, err := ca.MakeAndWriteSubCA(certFile, keyFile, serialFile, name, expireDays)
-// 	return subCA, true, err
-// }
-
-func (ca *CA) EnsureServerCertificate(certFile, keyFile string, hostnames []string, expireDays int) (*TLSCertificateConfig, bool, error) {
```

Comment:
just my 2c this naming is more idiomatic, the short name makes it more straightforward to reason about that this is a method off ca. where caClient feels like an internal object.

Reply:
I would prefer that it is not named CA so that someone who comes and tries to modify this code later one has one more warning that any CA functions (f.e. if you need to add CRLs, explicit renew(), etc) do not belong here. It is a wrapper. Client. The CA is in the backend which is pluggable and has multiple possible implementations.